### PR TITLE
Support the getrandom system call if glibc doesn't provide a wrapper

### PR DESCRIPTION
**Pull request by kroeckx**
Glibc only provides getrandom() since version 2.25, while the Linux
kernel has support for it since version 3.17.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/rand/rand_unix.c [hunk 1]
```diff
@@ -119,26 +126,52 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 #   error "Seeding uses urandom but DEVRANDOM is not configured"
 #  endif
 
+#  if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
```
Comment (richsalz) [inline crypto/rand/rand_unix.c]: This line and the next can be merged into a single three-part conditional.

**Comment by kroeckx** [comment]
Is someone able to test this on FreeBSD or OpenBSD? I don't have access to it. FreeBSD only added this like 2 weeks ago, so you need something very recent.

**Comment by richsalz** [comment]
And what happens on a FreeBSD system that is older than two weeks?

**Comment by kroeckx** [comment]
It falls back to trying /dev/urandom ...

**Comment by kroeckx** [comment]
FreeBSD also has a sysctl MIB (CTL_KERN.KERN_ARND), but it seems undocumented, so I'm not sure how to properly support it.

**Comment by kroeckx** [comment]
It seems to be documented in the random(4) manpage since FreeBSD 11.0

**Comment by kroeckx** [comment]
So I added something that should try to use a syscall on FreeBSD > 11. It would be nice if someone can test that on both FreeBSD 11 and 12.

**Comment by kroeckx** [comment]
@kaduk: Is this something you can test on FreeBSD?

File: crypto/rand/rand_unix.c [hunk 2]
```diff
@@ -119,26 +133,92 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+#   if __GLIBC_PREREQ(2, 25)
+#    define OPENSSL_HAVE_GETRANDOM
+#   endif
+#  endif
+
+#  if (defined(__FreeBSD__) && __FreeBSD__ >= 12)
```
Comment (kaduk) [inline crypto/rand/rand_unix.c]: It's probably better to use ```__FreeBSD_version``` from ```sys/param.h```and check the actual value against a fine-grained constant, since people might be running a snapshot of 12-current that is too old to have this support.  Unfortunately, the addition of getrandom does not seem to appear in https://www.freebsd.org/doc/en/books/porters-handbook/versions.html so some detective work will be required...

File: crypto/rand/rand_unix.c [hunk 3]
```diff
@@ -119,26 +133,92 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+#  if (defined(__FreeBSD__) && __FreeBSD__ >= 11)
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
+#   endif
+    return (int)sysctl_random(buf, buflen);
+#  endif
+   /* Supported since OpenBSD 5.6 */
+#  if defined(__OpenBSD__) && OpenBSD >= 201411
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
```
Comment (kaduk) [inline crypto/rand/rand_unix.c]: It feels pretty weird to be conditionally ```#define```ing ```OPENSSL_RAND_SEED_GETRANDOM``` in the body of a function like this.

**Comment by kroeckx** [comment]
I would like you to check that the system calls happen and doesn't fall back to /dev/urandom. Something like strace apps/openssl rand 1

File: crypto/rand/rand_unix.c (hunk 3 above)
Comment (kroeckx) [reply crypto/rand/rand_unix.c]: I didn't like it and had no better idea at that time, but I have now, let me add an other commit.

File: crypto/rand/rand_unix.c (hunk 2 above)
Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Oh, I didn't know 12 was already that old. I guess I could go for 1200061 which is the first version there after it's been merged.

Comment (kaduk) [reply crypto/rand/rand_unix.c]: Yeah, something that calls itself 12 has existed since 11.0 was released.

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Pushed a commit for that. Any idea when the sysctl support was added so I can do the same for that?

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Current oldest reference I could find is:
https://lists.freebsd.org/pipermail/svn-src-all/2012-July/056019.html

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: I seem to have found it: https://github.com/freebsd/freebsd/commit/3ef9d41401d2f9d9b32d2c476de4aee9b8049188

File: crypto/rand/rand_unix.c [hunk 4]
```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+#   define OPENSSL_HAVE_GETRANDOM
+#  endif
+
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+#   include <sys/random.h>
+#  endif
+
 #  if defined(OPENSSL_RAND_SEED_OS)
 #   if !defined(DEVRANDOM)
 #    error "OS seeding requires DEVRANDOM to be configured"
```
Comment (kaduk) [inline crypto/rand/rand_unix.c]: Is this actually true now, though?

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: It's used as fallback if the kernel doesn't have a system call. But it's true that if a system call is available, that it's no longer required.

File: crypto/rand/rand_unix.c [hunk 5]
```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+
+    mib[0] = CTL_KERN;
+    mib[1] = KERN_ARND;
+
+    do {
+        len = buflen;
+        if (sysctl(mib, 2, buf, &len, NULL, 0) == -1)
+            return done;
+        done += len;
+        buf += len;
```
Comment (kaduk) [inline crypto/rand/rand_unix.c]: arithmetic on a void pointer is a GNU extension
(and so at least the strict-warnings build fails.  "Who does non-strict-warnings builds anyway?")

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: I didn't even compile this ...

File: crypto/rand/rand_unix.c [hunk 6]
```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+    /* The sysctl returns longs. */
+    if (buflen % sizeof(long) != 0)
+        return 0;
+
+    mib[0] = CTL_KERN;
+    mib[1] = KERN_ARND;
+
+    do {
+        len = buflen;
+        if (sysctl(mib, 2, buf, &len, NULL, 0) == -1)
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply crypto/rand/rand_unix.c]: It's malloc()'d. I'm not sure why you think there is an alignment issue, or how I need to resolve it.

File: crypto/rand/rand_unix.c (hunk 5 above)
Comment (kaduk) [reply crypto/rand/rand_unix.c]: This was more about "I have no idea whether a non-strict-warnings build would succeed; I did not expect you to have built it on FreeBSD already.

File: crypto/rand/rand_unix.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (kaduk) [reply crypto/rand/rand_unix.c]: Well, a few lines up it's claimed that the data is returned in multiples of 'long', as if the long type was used internally.  The compiler will not complain at us if we pass a ```void*``` where a ```long*``` is expected, leaving it to the programmer to ensure that the pointer value in question has sufficient alignment for the type in question (since formally, we're supposed to only actually be using a pointer to the type that we end up dereferencing it as).  As you note, this buffer is going to start its life as a malloc output, which is not intrinsically type long.  That said, malloc output is required to be sufficiently aligned for use as any native type, so there should not be a problem in practice at runtime.  (There would maybe be issues if this function was exposed to user-supplied argument values, but I did not attempt to trace if that is possible.)

**Comment by kaduk** [comment]
It turns out that the truss(1) output is not terribly useful for the sysctl case, since so much information is hidden behind pointers that it's tough to be sure that the sysctl in the trace is actually the sysctl we're interested in.  That said, a well-placed fprintf() does confirm that the sysctl version runs successfully.

Interestingly, on my ```FreeBSD 12.0-CURRENT #1 r324846: Sun Oct 22 13:40:14 CDT 2017```, I can successfully configure and build ```--with-rand-seed=getrandom``` and no build-time error.  ```OPENSSL_HAVE_GETRANDOM``` ends up not being defined, and ```syscall_random()``` falls through to ```sysctl_random()```.  This seems a surprising behavior to me; it would be nice if we could arrange for the build to fail if getrandom is requested but not available.  (And should we be trying to use sysctl when the 'os' or default seeding is not in use?)

File: crypto/rand/rand_unix.c [hunk 7]
```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+ * returns the number of bytes returned in buf, or <= 0 on error.
+ */
+int syscall_random(void *buf, size_t buflen)
+{
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+    return (int)getrandom(buf, buflen, 0);
+#  endif
+
+#  if defined(__linux) && defined(SYS_getrandom)
+    return (int)syscall(SYS_getrandom, buf, buflen, 0);
```
Comment (mspncp) [inline crypto/rand/rand_unix.c]: I'm only a layman in this matter, so please excuse me if my question is off-topic or stupid: In this [LWN Article](https://lwn.net/Articles/711013) the author Jonathan Corbet describes that there was a lot of discussion about whether the glibc getrandom() call should be a thread cancellation point or not. Finally it was decided that it should be.

```
The most extensive argument, though, was over whether getrandom() should be a thread cancellation point.
In other words, what should happen if pthread_cancel() is called on a thread that is currently blocked
in getrandom()? The original patch did make getrandom() into a cancellation point; it still behaves
that way in the version merged for 2.25, but it had to survive a lot of argument to get there.
```

While the glibc getrandom() is a thread cancellation point, your raw syscall is not. Does this difference have any practical relevance for OpenSSL?

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: I don't think it's relevant. Cancellation points are just a bunch of functions where the cancellation can happen, which are typically function that can block.

**Comment by kroeckx** [comment]
I was planning on updating some comment, let me do that now.

**Comment by kroeckx** [comment]
I intend to squash this all into 1 commit and change the commit message to:
Add support for getting entropy using system calls

**Comment by kroeckx** [comment]
I had to resolve conflicts, I've fixed them and squashed all the commits.

File: crypto/rand/rand_unix.c [hunk 8]
```diff
@@ -149,26 +163,94 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 #   error "Seeding uses urandom but DEVRANDOM is not configured"
 #  endif
 
+#  if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 25)
+#    define OPENSSL_HAVE_GETRANDOM
+#   endif
+#  endif
+
+#  if (defined(__FreeBSD__) && __FreeBSD_version >= 1200061)
```
Comment (paulidale) [inline crypto/rand/rand_unix.c]: Worth splitting this like was done for `__GLIBC_PREREQ` ?

If neither `__FreeBSD__` nor `__FreeBSD_version` is defined, isn't there is a syntax error?

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: The reason for the __GLIBC_PREREQ() on the next line is that it takes parameters, so you need to be sure it exist before you use it.

File: crypto/rand/rand_unix.c [hunk 9]
```diff
@@ -149,26 +163,94 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
...
+    size_t len;
+
+    /*
+     * Old implementations returned longs, newer versions support variable
+     * sizes up to 256 byte. The code below would not work properly when
+     * the sysctl returns long and we want to request something not a multiple
+     * of longs.
+     */
+    if (buflen % sizeof(long) != 0)
+        return 0;
```
Comment (dot-asm) [inline crypto/rand/rand_unix.c]: Is there reason to punish caller that hard? Why not read buflen - buflen % sizeof(long), and then read one long to temporary storage and and copy buflen %sizeof(long) bytes?

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: The reason is that the loop is the only example I could find for something that is undocumented, and the comment is based on what I saw in the source, and I'm not even 100% sure that's correct. But that loop clearly has a problem if the implementation is really using a long.

I think us trying to detect which implementation is in the kernel, and then starting using longs is the wrong way to go. I have no idea if there are other cases where it returns a buffer that's shorter than the one we ask. But the code with longs didn't look at the length and always filled a whole long.

So I think the easiest way to deal with this is have the calling code always request a multiple of the size of a long, and throw away the rest if it needs less. We currently always have a multiple of the size of a long anyway.

**Comment by mspncp** [comment]
Thanks for testing! BTW: `--with-rand-seed=getrandom,os` can be omitted, the default value is `os` which on linux uses all reasonable random sources (getrandom first, then /dev/*random, etc.).

ping @openssl/committers, can anyone of you review this, so we can merge it before the next release?

File: INSTALL [hunk 10]
```diff
@@ -224,7 +224,8 @@
                    os:         Use a trusted operating system entropy source.
                                This is the default method if such an entropy
                                source exists.
-                   getrandom:  Use the L<getrandom(2)> system call if available.
```
Comment (mspncp) [inline INSTALL]: > equivalant -> equivalent

File: crypto/rand/rand_unix.c (hunk 5 above)
Comment (mspncp) [reply crypto/rand/rand_unix.c]: Any progress on FreeBSD?

**Comment by kroeckx** [comment]
We use getrandom() from glibc, not getentropy. And it's also not glibc specific. How about:
Add support for getting entropy from the OS without using files

FreeBSD was tested, but only the version that doesn't use getrandom(), which is was most people currently will end up using anyway.

**Comment by kroeckx** [comment]
Or just: Add support for getrandom() or equivalent system calls and use them by default

File: crypto/rand/rand_unix.c (hunk 9 above)
Comment (dot-asm) [reply crypto/rand/rand_unix.c]: > We currently always have a multiple of the size of a long anyway.

Then it would arguably be appropriate to use assert here.

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Is that ossl_assert()?

Comment (dot-asm) [reply crypto/rand/rand_unix.c]: In this case yes.

**Comment by kroeckx** [comment]
So can someone review this?

**Comment by mspncp** [comment]
If nobody else shows up, you can merge with my approval. But I'd really prefer if some else would have a look at this who is more familiar with the technical details.

File: crypto/rand/rand_unix.c (hunk 9 above)
Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Oh, I didn't know it could return a value. I just merged this like an hour ago. I'll make a new PR.

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: #6044

File: crypto/rand/rand_unix.c (hunk 5 above)
Comment (kaduk) [reply crypto/rand/rand_unix.c]: I finally got to install a FreeBSD 12-current VM and test a recent OpenSSL on it.
The test suite runs clean and I confirmed with gdb that we are using the getrandom() function from libc and it is returning (at least nominally) random data into the buffer.

**Comment by lukastribus** [comment]
@kroeckx when hitting this code path, `openssl version -r` still shows `Seeding source: os-specific` instead of getrandom(), but the library does use getrandom().

Can you take a look at that and make sure this ouput is consistent with the actual behavior?

**Comment by andreasschulze** [comment]
works as expected here: $ openssl11 version -a
OpenSSL 1.1.1-pre6 (beta) 1 May 2018
built on: Tue May  8 16:51:33 2018 UTC
platform: debian-amd64
options:  bn(64,64) rc4(16x,int) des(int) idea(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -g -O2 -fdebug-prefix-map=/build/dv-openssl-9iOTHP/dv-openssl-1.1.1~pre6=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wa,--noexecstack -Wall -g -O2 -fdebug-prefix-map=/build/dv-openssl-9iOTHP/dv-openssl-1.1.1~pre6=. -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
OPENSSLDIR: "/etc/ssl"
ENGINESDIR: "/usr/lib/x86_64-linux-gnu/engines-1.1"
Seeding source: getrandom-syscall os-specific

**Comment by mspncp** [comment]
> openssl version -r still shows Seeding source: os-specific instead of getrandom(), but the library does use getrandom().
>
> Can you take a look at that and make sure this ouput is consistent with the actual behavior?

The `os-specific` option is a bit more generic than options like `getrandom-syscall`; it means: "use the optimal source depending on the operating system". The current os default for linux is to try 1) the glibc getrandom() call, 2)  a raw getrandom() syscall, or 3) read from /dev/Xrandom, in this order and provided the respective method is available (AFAIR). So yes, it is expected that getrandom() is used if it exists and if `os-specific` is configured (which is the default).

**Comment by lukastribus** [comment]
So basically `Seeding source` is not what OpenSSL actually uses as entropy source, instead it is about the exact content of `--with-rand-seed` in the configure phase (or it's default: os).

And in the `--with-rand-seed` command, os is basically equivalent to getrandom,devrandom (at least on Linux and BSD).

**Comment by mspncp** [comment]
Correct.

**Comment by mspncp** [comment]
See https://github.com/openssl/openssl/blob/master/apps/version.c#L158
